### ISSUE TITLE
Stabilizing the Kumaraswamy Distribution

### ISSUE THREAD

**Issue by maxwass** [pytorch/pytorch#139019]
### 🐛 Describe the bug

The current implementation of the Kumaraswamy distribution has significant numerical instabilities in its log-pdf, inverse CDF, and their parameter gradient expressions. All such expressions contain unstable terms of the form `log(1 - exp(x))`. I believe these instabilities have significantly hindered the usage of the Kumarswamy.

I recently published a paper [Stabilizing the Kumaraswamy Distribution](https://arxiv.org/abs/2410.00660) which diagnoses these issues and provides a stable implementation with an unconstrained logarithmic parameterization.

My implementation inherits from `Distribution`, not `TransformedDistribution`; relying on the sequence of provided transformations applied to the base distribution, namely 

```
transforms = [
            PowerTransform(exponent=self.concentration0.reciprocal()),
            AffineTransform(loc=1.0, scale=-1.0),
            PowerTransform(exponent=self.concentration1.reciprocal()),
        ]

```
produces the aforementioned `log(1-exp(x))` instabilities.

If there is interest in correcting the PyTorch implementation, I am happy to help, I just ask the paper be cited in the documentation.

Let me know how we should proceed.

-------

Some Evidence of the communities struggle with these instabilities
- [torch.distributions.kumaraswamy.Kumaraswamy generates samples outside its support (0,1)](https://github.com/pytorch/pytorch/issues/95548)
- [Implement Kumaraswamy Distribution](https://github.com/probtorch/pytorch/pull/143)

See also Section 5: Related Work, 'Kumaraswamy as a Beta surrogate' in my paper. It briefly goes over the hacks required in prior works using the Kumaraswamy (across Pytorch, Tensorflow, Theano, etc), typically relying on clipping of KS parameters (thus limiting expressiveness of the distribution class) and/or constraining the standard uniform base distribution (and thus distorting the sampling distribution).

cc @fritzo @neerajprad @alicanb @nikitaved

### Versions

python collect_env.py
--2024-10-27 18:06:10--  https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 24366 (24K) [text/plain]
Saving to: ‘collect_env.py’

collect_env.py                   100%[========================================================>]  23.79K  --.-KB/s    in 0.03s   

2024-10-27 18:06:10 (810 KB/s) - ‘collect_env.py’ saved [24366/24366]

Collecting environment information...
PyTorch version: 2.4.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.19 | packaged by conda-forge | (main, Mar 20 2024, 12:55:20)  [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-13.0.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.11.0
[pip3] pytorch-lightning==2.2.2
[pip3] torch==2.4.0
[pip3] torch_geometric==2.5.3
[pip3] torchmetrics==1.4.0
[pip3] torchrl==0.5.0
[pip3] torchvision==0.16.1
[conda] libtorch                  2.1.2           cpu_generic_h811e168_4    conda-forge
[conda] nomkl                     1.0                  h5ca1d4c_0    conda-forge
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] optree                    0.11.0                   pypi_0    pypi
[conda] pytorch-lightning         2.2.2              pyhd8ed1ab_0    conda-forge
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torch-geometric           2.5.3                    pypi_0    pypi
[conda] torchmetrics              1.4.0              pyhd8ed1ab_0    conda-forge
[conda] torchrl                   0.5.0                    pypi_0    pypi
[conda] torchvision               0.16.1          cpu_py39h936d113_3    conda-forge

**Comment by vmoens** [comment]
Hello @maxwass 
That seems interesting!
If you want to make a PR with it we'd love to review it. 

The basic requirements are that all the tests should pass for that dist, you can find them there:
https://github.com/pytorch/pytorch/blob/9af18169742c7813dcfcb7497525d2df6644c0a7/test/distributions/test_distributions.py#L3584

> My implementation inherits from Distribution, not TransformedDistribution

That could be an issue though because a chunk of the user base might do things like
```python
from torch.distributions import Kumaraswamy
dist = Kumaraswamy(**kwargs)
t0 = dist.transform[0]
# do something with t0
```

But we can find solutions to that. One could be to pass a `stable` kwarg and actually build another class if this is passed, here is an example:
```python
class _KumaraswamyMeta(abc.ABCMeta):
    def __call__(cls, *args, **kwargs):
        if kwargs.pop("stable", False):
            return something_else(*args, **kwargs)
        return super().__call__(*args, **kwargs)

class Kumaraswamy(Distribution, metaclass=_KumaraswamyMeta):
    ...
```

cc @fritzo

**Comment by maxwass** [comment]
My stable reparameterization is in terms of the unconstrained log parameters (which is a big advantage), while the current PyTorch implementation uses positive parameters. 

This poses two issues in terms of passing the linked tests.

1. The names of my arguments are log_concentration0/1, breaking some of the tests.


2. The mean and variance test in
https://github.com/pytorch/pytorch/blob/9af18169742c7813dcfcb7497525d2df6644c0a7/test/distributions/test_distributions.py#L3584
will be have their parameters exponentiated and so give different answers.

**Comment by vmoens** [comment]
We need these behaviours to match exactly, unfortunately. We could add a keyword argument to make it possible to pass unconstrained arguments (like we do with categorical), but we cannot make the BC-breaking change of using unbounded (log) params when they were not used before without warning.

**Comment by maxwass** [comment]
Understood.

Would it perhaps be easier to have a new distribution, perhaps called `KumaraswamyStable`, with unconstrained log params and the stable log-pdf/inverse CDF implementation?

**Comment by maxwass** [comment]
Bump @vmoens @fritzo.
